[DNM] Unifying H4 UART transport for host and controller

nordic-krch opened the pull request:
This is an outcome of the discussion in #26445 and #26437 where support for asynchronous UART API was added to host and controller. This created some redundancy because handling of UART data was done independently in hci_uart sample and h4 module (used by the host). This PR attempts to unify handling of UART data on host and controller by creating transport library: `h4_uart`.

`h4_uart` transport supports both UART APIs: interrupt driven and asynchronous.

As the outcome of adding `h4_uart` following changes were done:
 - H4 was adapted to use `h4_uart`.
 - hci_uart sample was adopted to use `h4_uart`
- added nrf9160dk_nrf9160 configuration for `peripheral_hr` sample which is using asynchronous UART API
- added nrf9160dk_nrf52840 configuration for `hci_uart` sample which is using asynchronous UART API (previously `hci_uart_async` sample was created)

Fixes #2955.

nordic-krch committed:
drivers: bluetooth: hci: Add h4_uart transport

Added h4 uart transport which can be used by host and controller.
Transport supports 2 UART APIs: interrupt driven and asynchronous.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: bluetooth: hci: Update H4 to use h4_uart transport

Removed UART handling part from H4 and replaced it with h4_uart
transport which handles data flow over UART.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
samples: bluetooth: peripheral_hr: Add nrf9160dk_nrf9160 board

Added configuration for nrf9160dk_nrf9160 board. Configuration
is using asynchronous UART API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
samples: bluetooth: hci_uart: Modify sample to use h4_uart

Modified sample to use h4_uart transport.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
samples: bluetooth: hci_uart: Update nrf9160dk_nrf52840 configuration

Updated configuration to use asynchronous UART API.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

jhedberg commented on drivers/bluetooth/hci/h4_uart.c:
```diff
@@ -0,0 +1,563 @@
+	};
+
+	for (i = 0; i < sizeof(cc_evt); i++) {
+			uart_poll_out(transport->dev,
```
Incorrect indentation here (I'm surprised checkpatch didn't catch this)

carlescufi commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -311,46 +297,12 @@ void main(void)
-		if (err) {
-			LOG_ERR("Failed to send");
-		}
+		err = h4_uart_write(&transport,
```
I think you are fixing:

https://github.com/zephyrproject-rtos/zephyr/issues/2955

nordic-krch replied:
yes

jhedberg commented on drivers/bluetooth/hci/h4_uart.c:
```diff
@@ -0,0 +1,563 @@
+		}
+	}
+
+#if CONFIG_BT_H4_UART_INTERRUPT_DRIVEN
+	uart_irq_tx_enable(transport->dev);
+#else
+	int err = send(transport);
+	if (err < 0) {
+		on_tx_buf_complete(&transport->tx);
+		return err;
+	}
+#endif
```
`IS_ENABLED()`?

jhedberg commented on drivers/bluetooth/hci/h4_uart.c:
```diff
@@ -0,0 +1,563 @@
+int h4_uart_write(struct h4_uart *transport, struct net_buf *buf)
+{
+	if (!atomic_ptr_cas((atomic_ptr_t *)&transport->tx.curr,
+				NULL, buf)) {
```
This doesn't look like it's properly aligned with the ( of the first line

jhedberg commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -48,197 +38,180 @@ static K_FIFO_DEFINE(tx_queue);
-			continue;
+	if (rx.have_hdr) {
+		rx.buf = bt_buf_get_tx(BT_BUF_H4, K_FOREVER, &rx.type,
+					    sizeof(rx.type));
```
Alignment

jhedberg commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -48,197 +38,180 @@ static K_FIFO_DEFINE(tx_queue);
-	while (buf->len) {
-		uart_poll_out(hci_uart_dev, net_buf_pull_u8(buf));
+	LOG_DBG("remaining %u discard %u have_hdr %u rx.buf %p len %u",
+	       rx.remaining, rx.discard, rx.have_hdr, rx.buf,
+	       rx.buf ? rx.buf->len : 0);
```
Doesn't look correctly aligned

carlescufi commented on drivers/bluetooth/hci/Kconfig:
```diff
@@ -71,6 +76,50 @@ config BT_NO_DRIVER
+choice
+	prompt "UART API"
+
+config BT_H4_UART_INTERRUPT_DRIVEN
```
Why do we need this config option? Have we ever supported h4 on a UART that only has poll_in and poll_out?

carlescufi replied:
Oh, I see, it's a choice, sorry.

jhedberg reviewed (changes requested):
Just a quick review of some coding style related issues I spotted. I hope to have more time do do a thorough review later.

carlescufi reviewed (changes requested):
Thanks! A few comments

carlescufi commented:
This needs testing on at least:

- QEMU + hci_uart on nRF DK
- native_posix + hci_uart on nRF DK
- 52 DK + hci_uart on 52 DK
- 91 DK

carlescufi commented:
@nordic-krch @anangl recently fixed a bug in the h4.c driver (https://github.com/zephyrproject-rtos/zephyr/pull/26563). I think it doesn't apply anymore in the new, but can we double-check?

carlescufi commented:
@jhedberg and @joerc I'd be really interested to get your perspectives on #26445 + #26437 vs this pull request. I personally like the unification, but it does come at a cost (the extra buffers, and another layer on top of the UART driver).

nordic-krch commented on boards/arm/nrf9160dk_nrf9160/nrf52840_reset.c:
```diff
@@ -50,11 +49,13 @@ int bt_hci_transport_setup(struct device *h4)
 	 */
 	k_sleep(K_MSEC(10));
 
+#ifdef CONFIG_BT_H4_UART_INTERRUPT_DRIVEN
```
i'm a biggest fun of `IS_ENABLED()` but here presence of API is ifdefed so IS_ENABLED cannot be used.

jhedberg replied:
Where is the API located? We shouldn't have #ifdefs around function declarations in header files, so if that's the problem it needs to be fixed.

nordic-krch replied:
totally agree:
https://github.com/zephyrproject-rtos/zephyr/blob/1886cb4b4e6695b5cd48373dd7c4737857d4f282/include/drivers/uart.h#L703

nordic-krch replied:
i will create issue for that.

nordic-krch replied:
#26801

nordic-krch commented on drivers/bluetooth/hci/h4_uart.c:
```diff
@@ -0,0 +1,563 @@
+				K_NO_WAIT);
+	__ASSERT_NO_MSG(err == 0);
+
+#if CONFIG_BT_H4_UART_INTERRUPT_DRIVEN
+	uart_irq_rx_enable(transport->dev);
+#else
+	err = uart_rx_enable(transport->dev, buf, block_size, 1);
+	if (err < 0) {
+		return -EIO;
+	}
+#endif
```
😢 no

nordic-krch commented on drivers/bluetooth/Kconfig:
```diff
@@ -10,7 +10,7 @@
 # options are only applicable if controller support hasn't been enabled.
 menu "Bluetooth Drivers"
-	depends on BT && !BT_CTLR
+	depends on BT
```
what about `BT_HCI && (!BT_HCI_HOST || !BT_CTLR)`? I mean: if BT does not have controller or host.

joerchan replied:
@nordic-krch I think that makes sense, although i would find it more readable with `BT_HCI && !(BT_HCI_HOST && BT_CTLR)` i.e if BT does not have both controller and host.

nordic-krch commented on drivers/bluetooth/hci/h4_uart.c:
```diff
@@ -0,0 +1,563 @@
+
+	len = MIN(req_len, data->len - data->offset);
+	if (dst) {
+		memcpy(dst, &data->buf[data->offset], len);
```
yes, I will try to come up with some numbers regarding cpu load.

nordic-krch commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -1,45 +1,39 @@
 static K_THREAD_STACK_DEFINE(rx_thread_stack, CONFIG_BT_RX_STACK_SIZE);
-static struct k_thread rx_thread_data;
+
+static __aligned(sizeof(uint32_t)) uint8_t
```
yes, for async uart these buffers are required because uart use them directly. I'll check if in case of interrupt driven these can be fixed to some small number (will measure cpu load).

nordic-krch commented:
Please hold on review of `h4_uart.c`. Per offline discussion with @anangl i'm reworking buffer handling there to get rid of mem slabs and lists and use fixed 2 buffers which should be simpler.

nordic-krch committed:
[tmp] lib: os: ring_buffer: Allow using full buffer capacity

Previously, ring buffer had capacity of provided buffer size - 1. This
trick was used to distinguish between empty and full states. It had one
drawback: ring buffer could not be used as a pool of equal sized buffers
(using ring_buf_put_claim and ring_buf_get_claim).
Reworked internals to use non wrapping head and tail. Since they are
non wrapping, there is no issue with distinguishing between empty and
full. Since this appraoch would be vulnerable to wrapping on 32 bit
boundary, added a mechanism which periodically reduces all indexes to
avoid 32 bit wrapping.

After this rework, buffer has one byte more capacity. Simple test shows
slight performance improvement.

Updated tests to reflect increased capacity and added test to check if
it is possible to continuesly allocated 2 buffers of half ring buffer
size.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
[tmp] drivers: serial: nrf_uarte: Add detection of late rx_buf_rsp call

Added return of an error when uart_rx_buf_rsp is called too late.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
added DNM for now because it is based on few unmerged PRs in UART and ring_buffer.

joerchan commented:
@nashif The stale-bot seems to tag closed PRs as well.
